hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java

```diff
@@ -872,13 +873,14 @@ public void destroyAndCollectBlocks(ReclaimContext reclaimContext) {
     if (sf != null) {
       sf.clear(reclaimContext, this);
     }
-    for (INode child : getChildrenList(Snapshot.CURRENT_STATE_ID)) {
+
+    ReadOnlyList<INode> childs = getChildrenList(Snapshot.CURRENT_STATE_ID);
+    for (INode child : childs) {
       child.destroyAndCollectBlocks(reclaimContext);
     }
     if (getAclFeature() != null) {
       AclStorage.removeAclFeature(getAclFeature());
     }
-    clear();
```

Comment:
clear() try to set parent = null which will cause removeChilds() fail. Because the hierarchical info was destroyed.